Compatibility analyzer breaking change (#20567)
dotnet/docs: includes/core-changes/codeanalysis/5.0/ca1416-platform-compatibility-analyzer.md

### CA1416: Platform compatibility

.NET code analyzer rule CA1416 is enabled, by default, starting in .NET 5.0. It produces a build warning for calls to platform-specific APIs from call sites that don't verify the operating system.

#### Change description

Starting in .NET 5.0, the .NET SDK includes [.NET source code analyzers](../../../../docs/fundamentals/productivity/code-analysis.md). Several of these rules are enabled, by default, including CA1416. If your project contains code that violates this rule and is configured to treat warnings as errors, this change could break your build. Rule CA1416 informs you when you're using platform-specific APIs from places where the platform context isn't verified.

> Rule CA1416, the platform compatibility analyzer, works hand-in-hand with some other features that are new in .NET 5.0. .NET 5.0 introduces the <xref:System.Runtime.Versioning.MinimumOSPlatformAttribute> (soon to be `SupportedOSPlatformAttribute`) and <xref:System.Runtime.Versioning.RemovedInOSPlatformAttribute> (soon to be `UnsupportedOSPlatformAttribute`) attributes, which let you specify which platforms an API *is* or *isn't* supported on. In the absence of these attributes, an API is assumed to be supported on all platforms. These attributes have been applied to platform-specific APIs in the core .NET libraries.

Reviewer: Since the attributes have already been renamed in RC1, I suggest we invert the phrasing here to first mention the new name.

```suggestion
Rule CA1416, the platform compatibility analyzer, works hand-in-hand with some other features that are new in .NET 5.0. .NET 5.0 introduces `SupportedOSPlatformAttribute` and `UnsupportedOSPlatformAttribute` attributes (named <xref:System.Runtime.Versioning.MinimumOSPlatformAttribute> and <xref:System.Runtime.Versioning.RemovedInOSPlatformAttribute> in an earlier preview release), which let you specify which platforms an API *is* or *isn't* supported on. In the absence of these attributes, an API is assumed to be supported on all platforms. These attributes have been applied to platform-specific APIs in the core .NET libraries.
```

  ```csharp
  public void PlayCMajor()
  {
      Console.Beep(261, 1000);
  }
  ```

- The <xref:System.Drawing.Image.FromFile(System.String)?displayProperty=nameWithType> method is not supported in the browser (it's decorated with `[UnsupportedOSPlatform("browser")]`). The following code will produce a CA1416 warning at build time if the project uses the Blazor WebAssembly SDK (`<Project Sdk="Microsoft.NET.Sdk.BlazorWebAssembly">`) or includes `browser` as a supported platform (`<SupportedPlatform Include="browser" />`) in the project file.

  ```csharp
  public void CreateImage()
  {
      Image newImage = Image.FromFile("SampImag.jpg");
  }
  ```

#### Version introduced

5.0 RC1

#### Recommended action

Ensure that platform-specific APIs are only called when the code is running on an appropriate platform. You can achieve this using preprocessor directives, or at run time by using platform guards:

> - Add a `#if` [preprocessor directive](../../../../docs/csharp/language-reference/preprocessor-directives/preprocessor-if.md) around platform-specific API calls:

Reviewer: I believe the user would also need to _define_ this constant too, with something like the following.  Is that right, @dsplaisted?

```xml
<PropertyGroup>
	<DefineConstants Condition="'$(TargetFramework)' == 'net5.0-windows'">NET50_WINDOWS</DefineConstants>
</PropertyGroup>
```

Author: [This section](https://docs.microsoft.com/en-us/dotnet/standard/frameworks#how-to-specify-a-target-framework) says they're predefined, but I should add the .NET 5 symbols to the table (last part of the section).

Reviewer: Ah, cool; thanks. I'm not seeing this work on my machine, but I'm not sure why.

Author: @jeffhandley Can you try NET5_0-WINDOWS? I think I need to correct this snippet.

Reviewer: I've tried TFMs of `net5.0-windows`, `net5.0-windows7.0`, and `net5.0-windows10.0.19041`, and with each of them `DefineConstants` gets produced as `TRACE;DEBUG;NET;NET5_0`.

When trying to use `#if NET5_0-WINDOWS`, I also get an error that the `-` is an unexpected character.

Author: Probably the hyphen needs to be changed to an underscore too, so `NET5_0_WINDOWS`. But it sounds like that still wouldn't be valid...

Reviewer: Yeah, I tried that before checking what constants are getting generated.  I wonder if we should just remove that section until we get this detail worked out.

Reviewer: @gewarren Based on a meeting earlier this week related to what constants will get defined, we are not yet defining the platform-specific constants like this in the SDK, so I think we're good to leave this `#if def` section out permanently.

Author: Thanks for updating me.

### CA1416: Platform compatibility

.NET code analyzer rule CA1416 is enabled, by default, starting in .NET 5.0. It produces a build warning for calls to platform-specific APIs from call sites that don't verify the operating system.

#### Change description

Starting in .NET 5.0, the .NET SDK includes [.NET source code analyzers](../../../../docs/fundamentals/productivity/code-analysis.md). Several of these rules are enabled, by default, including CA1416. If your project contains code that violates this rule and is configured to treat warnings as errors, this change could break your build. Rule CA1416 informs you when you're using platform-specific APIs from places where the platform context isn't verified.

> Rule CA1416, the platform compatibility analyzer, works hand-in-hand with some other features that are new in .NET 5.0. .NET 5.0 introduces the <xref:System.Runtime.Versioning.OSPlatformAttribute> (soon to be `SupportedOSPlatformAttribute`) and <xref:System.Runtime.Versioning.RemovedInOSPlatformAttribute> (soon to be `UnsupportedOSPlatformAttribute`) attributes, which let you specify which platforms an API *is* or *isn't* supported on. In the absence of these attributes, an API is assumed to be supported on all platforms. These attributes have been applied to platform-specific APIs in the core .NET libraries.

Reviewer 2: `<xref:System.Runtime.Versioning.OSPlatformAttribute>` => `<xref:System.Runtime.Versioning.MinOSPlatformAttribute>`
```suggestion
Rule CA1416, the platform compatibility analyzer, works hand-in-hand with some other features that are new in .NET 5.0. .NET 5.0 introduces the <xref:System.Runtime.Versioning.MinOSPlatformAttribute> (soon to be `SupportedOSPlatformAttribute`) and <xref:System.Runtime.Versioning.RemovedInOSPlatformAttribute> (soon to be `UnsupportedOSPlatformAttribute`) attributes, which let you specify which platforms an API *is* or *isn't* supported on. In the absence of these attributes, an API is assumed to be supported on all platforms. These attributes have been applied to platform-specific APIs in the core .NET libraries.
```

In projects that target platforms for which APIs that they use aren't available, rule CA1416 flags any platform-specific API call where the platform context isn't verified. Most of the APIs that are now decorated with the `SupportedOSPlatformAttribute` and `UnsupportedOSPlatformAttribute` attributes throw <xref:System.PlatformNotSupportedException> exceptions when they're invoked on an unsupported operating system. Now that these APIs are marked as platform-specific, rule CA1416 helps you prevent run-time <xref:System.PlatformNotSupportedException> exceptions by adding OS checks to your call sites.

#### Examples

- The <xref:System.Console.Beep(System.Int32,System.Int32)?displayProperty=nameWithType> method is only supported on Windows (it's decorated with `[SupportedOSPlatform("windows")]`). The following code will produce a CA1416 warning at build time if the project [targets](../../../../docs/standard/frameworks.md) `net5.0` (but not `net5.0-windows`). For actions you can take to avoid the warning, see [Recommended action](#recommended-action).

  ```csharp
  public void PlayCMajor()
  {
      Console.Beep(261, 1000);
  }
  ```

> - The <xref:System.Drawing.Image.FromFile(System.String)?displayProperty=nameWithType> method is not supported in the browser (it's decorated with `[UnsupportedOSPlatform("browser")]`). The following code will produce a CA1416 warning at build time if the project uses the Blazor WebAssembly SDK (`<Project Sdk="Microsoft.NET.Blazor.WebAssembly.Sdk">`) or includes `browser` as a supported platform in the project file (`<SupportedPlatform Include="browser" />`).

Author: Not sure about `Project Sdk="Microsoft.NET.Blazor.WebAssembly.Sdk"`. Please correct!

Reviewer 2: Seems `Project Sdk="Microsoft.NET.Sdk.BlazorWebAssembly"`

> - The <xref:System.Drawing.Image.FromFile(System.String)?displayProperty=nameWithType> method is not supported in the browser (it's decorated with `[UnsupportedOSPlatform("browser")]`). The following code will produce a CA1416 warning at build time if the project uses the Blazor WebAssembly SDK (`<Project Sdk="Microsoft.NET.Blazor.WebAssembly.Sdk">`) or includes `browser` as a supported platform (`<SupportedPlatform Include="browser" />`) in the project file.

Reviewer 2: ```suggestion
- The <xref:System.Drawing.Image.FromFile(System.String)?displayProperty=nameWithType> method is not supported in the browser (it's decorated with `[UnsupportedOSPlatform("browser")]`). The following code will produce a CA1416 warning at build time if the project uses the Blazor WebAssembly SDK (`<Project Sdk="Microsoft.NET.Sdk.BlazorWebAssembly">`) or includes `browser` as a supported platform (`<SupportedPlatform Include="browser" />`) in the project file.
```

- Add a `#if` [preprocessor directive](../../../../docs/csharp/language-reference/preprocessor-directives/preprocessor-if.md) around platform-specific API calls:

  ```csharp
  #if NET50_WINDOWS
      Console.Beep(261, 1000);
  #endif
  ```

- Check the current operating system using one of the <xref:System.OperatingSystem?displayProperty=nameWithType> platform-checking methods, for example, `System.OperatingSystem.IsWindows()`, before calling a platform-specific API. You can use one of these methods in the condition of an `if` statement:

  ```csharp
  public void PlayCMajor()
  {
      if (OperatingSystem.IsWindows())
      {
          Console.Beep(261, 1000);
      }
  }
  ```

  ...or pass its result to <xref:System.Diagnostics.Debug.Assert(System.Boolean)?displayProperty=nameWithType>:

  ```csharp
  public void PlayCMajor()
  {
      Debug.Assert(OperatingSystem.IsWindows());
      Console.Beep(261, 1000);
  }

>   ```

Reviewer 3: Shouldn't the compiler omits the call to Debug.Assert in a release build and produces the warning for this? i.e. this snippet should fix the warning only for a debug build.

Do you know who can confirm?

I also cannot see IsWindows method in the current API reference. Is it missing or not shipped yet?

Author: IsWindows() is an RC1 method. We haven't updated the API docs yet for RC1.

For Debug.Assert(), I think the analyzer is smart enough to detect it, even for a release build. @buyaa-n can confirm.

Reviewer 2: > Shouldn't the compiler omits the call to Debug.Assert in a release build and produces the warning for this? i.e. this snippet should fix the warning only for a debug build.

Yes, it is trimmed for release build and that is the whole point, we don't want it fail runtime.  If dev doesn't want to add new if check and make a call to OperatingSystem.IsWindows() at runtime they could use it

Reviewer 2: @gewarren you might want to add this `Using Debug.Assert allows the check to be trimmed out of release builds`

Reviewer 2: Oh it was added below, never mind 😄

  Use <xref:System.Diagnostics.Debug.Assert(System.Boolean)?displayProperty=nameWithType> when you want the checks to be trimmed out of release builds.

You can also mark your API as platform-specific, in which case the burden of checking requirements falls on your callers. You can mark specific methods or types or an entire assembly.

```csharp
[SupportedOSPlatform("windows")]

> public void PlayCMajor ()

Reviewer 3: nit:

```suggestion
public void PlayCMajor()
```

{
    Console.Beep(261, 1000);
}
```

If you don't want to fix all your call sites, you can choose one of the following options to suppress the warning:

- To suppress rule CA1416, you can do so using `#pragma` or the [-nowarn](../../../../docs/csharp/language-reference/compiler-options/nowarn-compiler-option.md) compiler flag, or by [setting the rule's severity](../../../../docs/fundamentals/productivity/configure-code-analysis-rules.md#suppress-violations) to `none` in an .editorconfig file.

  ```csharp
  public void PlayCMajor ()
  {
  #pragma warning disable CA1416
      Console.Beep(261, 1000);
  #pragma warning restore CA1416
  }
  ```

- To disable code analysis completely, set `EnableNETAnalyzers` to `false` in your project file. For more information, see [EnableNETAnalyzers](../../../../docs/core/project-sdk/msbuild-props.md#enablenetanalyzers).

#### Category

- Code analysis
- Core .NET libraries

#### Affected APIs

For Windows platform:

> - All APIs listed at <https://github.com/dotnet/designs/blob/master/accepted/2020/windows-specific-apis/windows-specific-apis.md>.

Author: I am planning to list all the affected APIs out individually. LMK if that should be done before being published or if it's okay to come later.
